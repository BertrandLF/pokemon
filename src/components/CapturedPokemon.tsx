import React from 'react'
import './styles/CapturedPokemon.css';
import Pokemon from './Pokemon';

interface CapturedPokemonProps {
  pokemon: Pokemon;
  handleOnClick: (pokemon: Pokemon) => void;
}

class CapturedPokemon extends React.Component<CapturedPokemonProps> {

  constructor (props: Readonly<CapturedPokemonProps>){
    super(props);
    this.getGif = this.getGif.bind(this);
    this.setNormalizedSrc = this.setNormalizedSrc.bind(this);
  }

  pokeGifUrl: string = 'http://pokestadium.com/sprites/xy/';

  getGif(pokemon: string) {
    return this.pokeGifUrl + pokemon + '.gif';
  }

  setNormalizedSrc(img: any){
    const normalizedPokemonName: string = this.props.pokemon.name.replace('-', '');
    img.target.src = this.pokeGifUrl + normalizedPokemonName + '.gif';
  }

  render() {
    const pokemon = this.props.pokemon;
    return (
      <div className='captured-pokemon' onClick={() => this.props.handleOnClick(pokemon)}>
        <img onError={this.setNormalizedSrc} src={this.getGif(pokemon.name)} className='sprite-image' alt="sprite" />
        <div className="pokemon-data">
          <div className="data-row">
            <p className="cell">ID:</p>
            <p className="cell">{pokemon.id}</p>
          </div>
          <div className="data-row">
            <p className="cell">Name:</p>
            <p className="cell">{pokemon.name}</p>
          </div>
          <div className="data-row">
            <p className="cell">Type:</p>
            <p className="cell">{pokemon.type}</p>
          </div>
        </div>
      </div>
    )
  }
}

export default CapturedPokemon;