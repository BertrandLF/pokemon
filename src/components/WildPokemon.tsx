import React from 'react';
import './styles/WildPokemon.css';
import Pokemon from './Pokemon';

interface WildPokemonProps {
  pokemon: Pokemon;
  handleOnClick: (id: number) => void;
}

class WildPokemon extends React.Component<WildPokemonProps> {

  render() {
    const pokemon = this.props.pokemon;
    const style = { backgroundImage: `url(${pokemon.sprite})` };
    return <button
      className="wild-pokemon"
      style={style}
      onClick={() => this.props.handleOnClick(pokemon.id)} >
    </button>
  }

}

export default WildPokemon;