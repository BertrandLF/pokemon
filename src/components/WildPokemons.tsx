import React from 'react';
import './styles/WildPokemons.css';
import PokeCell from './PokeCell';
import Pokemon from './Pokemon';

interface WildPokemonsProps {
  pokemons: Set<Pokemon>;
  handleOnClick: (id: number) => void;
}
class WildPokemons extends React.Component<WildPokemonsProps> {

  cells() {
    return Array.from(this.props.pokemons.values()).map((pokemon: Pokemon) =>
      <PokeCell
        key={pokemon.id}
        pokemon={pokemon}
        handleOnClick={this.props.handleOnClick}
      />
    );
  }

  render() {
    return (
      <section className="wild-pokemons">
        {this.cells()}
      </section>
    )
  }
}

export default WildPokemons;