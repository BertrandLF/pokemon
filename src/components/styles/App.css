.App {
  display: flex;
  justify-content: center;
  width: 63em;
  height: 50em;
  margin: 3em auto;
  border-radius: 1em;
  box-shadow: 0 1em 4em rgba(0, 0, 0, 0.22), 0 1em 4em rgba(0, 0, 0, 0.22);
}