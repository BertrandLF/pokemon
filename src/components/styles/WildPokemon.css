.wild-pokemon {
  width: 10em;
  height: 9em;
  margin: 0.7em;
  background-color:rgb(145, 75, 177);
  background-repeat: no-repeat;
  border: none;
}