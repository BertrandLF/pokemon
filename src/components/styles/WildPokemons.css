.wild-pokemons {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  align-items: center;
  width: 50%;
  height: 100%;
  margin-right: 0.7em;
  background-color: rgb(145, 75, 177);
  border-top-left-radius: 1em;
  overflow: scroll;
}