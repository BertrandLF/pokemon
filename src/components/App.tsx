import React, { useState, useEffect } from 'react';
import WildPokemons from './WildPokemons';
import CapturedPokemons from './CapturedPokemons';
import Pokemon from './Pokemon';
import './styles/App.css';

const App = () => {

  const [wildPokemons, setWildPokemons] = useState(new Map<number, Pokemon>());
  const [capturedPokemons, setCapturedPokemons] = useState(new Map<number, Pokemon>());
  const [pokeBalls, setPokeBalls] = useState(6);
  const [loadedPokemonId, setLoadedPokemonId] = useState(1);

  useEffect(() => {
    const addPokemon = () => {
      if (loadedPokemonId < 152) {
        fetchPokemon(loadedPokemonId)
          .then(data => {
            const pokemon = new Pokemon(data);
            wildPokemons.set(pokemon.id, pokemon);
            setWildPokemons(wildPokemons);
            setLoadedPokemonId(loadedPokemonId + 1);
          })
          .catch(err => console.log('Failed to load pokemon!', err));
      }
    }
    var timerID = setInterval(() => addPokemon(), 500);

    return function cleanup() {
      clearInterval(timerID);
    };
  }, [wildPokemons, loadedPokemonId]);

  const capture = (id: number) => {
    if (pokeBalls > 0) {
          const pokemon = wildPokemons.get(id);
          if (pokemon) {
            capturedPokemons.set(pokemon.id, pokemon);
            wildPokemons.delete(pokemon.id);
            setCapturedPokemons(capturedPokemons);
            setWildPokemons(wildPokemons);
            setPokeBalls(pokeBalls - 1);
          } else {
            console.log('Failed to catch it!');
          }
    }
  }

  const fetchPokemon = (id: number): Promise<JSON> => {
    return fetch(`https://pokeapi.co/api/v2/pokemon/${id}`, {
      method: "GET",
      headers: {
        "access-control-allow-origin": "*",
        "Content-type": "application/json; charset=UTF-8"
      }
    })
      .then(res => res.json());
  }

  const release = (pokemon: Pokemon) => {
    capturedPokemons.delete(pokemon.id);
    wildPokemons.set(pokemon.id, pokemon);
    setCapturedPokemons(capturedPokemons);
    setWildPokemons(wildPokemons);
    setPokeBalls(pokeBalls + 1);
  }

  return (
    <div className="App">
      <WildPokemons
        pokemons={wildPokemons}
        handleOnClick={capture} />
      <CapturedPokemons
        pokemons={capturedPokemons}
        handleOnClick={release} />
    </div>
  );

}

export default App;
