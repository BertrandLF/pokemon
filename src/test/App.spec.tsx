import React from 'react';
import App from '../components/App';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReactTestRenderer, { act } from 'react-test-renderer';
import retry from '@skidding/async-retry';
import fetchMock from 'fetch-mock';
import { psyduckData, psyduck } from './PokemonFixture';
import WildPokemon from '../components/WildPokemon';
import WildPokemons from '../components/WildPokemons';
import CapturedPokemon from '../components/CapturedPokemon';

it('renders without crashing', () => {
  const app = shallow(<App />);
  expect(toJson(app)).toMatchSnapshot()
});

describe('App effects tests', () => {

  fetchMock.get('begin:https://pokeapi.co/api/v2/pokemon', psyduckData);
  let container: any;
  let containerInstance: any;
  beforeEach(() => {
    act(() => {
      container = ReactTestRenderer.create(<App />)
    });
    containerInstance = container.root;
  })

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('adds pokemons automatically', async () => {
    const expectedWildPokemons = new Map([[psyduck.id, psyduck]]);
    const wildPokemons = containerInstance.findByType(WildPokemons);

    await retry(() => {
      expect(wildPokemons.props.pokemons).toEqual(expectedWildPokemons);
    });
  });

  it('captures pokemons', async () => {
    let wildPokemon: any;

    await retry(() => {
      wildPokemon = containerInstance.findByType(WildPokemon);
      expect(wildPokemon.props.pokemon).toEqual(psyduck);
    });

    act(() => {
      wildPokemon.props.handleOnClick(psyduck.id);
    });

    await retry(() => {
      const capturedPokemon = containerInstance.findByType(CapturedPokemon);
      expect(capturedPokemon.props.pokemon).toEqual(psyduck);
    });

  });

  it('releases pokemons', async() => {
    let wildPokemon: any;
    let capturedPokemon: any;

    await retry(() => {
      wildPokemon = containerInstance.findByType(WildPokemon);
      expect(wildPokemon.props.pokemon).toEqual(psyduck);
    });

    act(() => {
      wildPokemon.props.handleOnClick(psyduck.id);
    });

    await retry(() => {
      capturedPokemon = containerInstance.findByType(CapturedPokemon);
      expect(capturedPokemon.props.pokemon).toEqual(psyduck);
    });

    act(() => {
      capturedPokemon.props.handleOnClick(psyduck);
    })

    await retry(() => {
      wildPokemon = containerInstance.findByType(WildPokemon);
      expect(wildPokemon.props.pokemon).toEqual(psyduck);
    });

  });



})
