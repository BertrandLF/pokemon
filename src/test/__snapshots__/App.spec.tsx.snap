// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="App"
>
  <WildPokemons
    handleOnClick={[Function]}
    pokemons={Map {}}
  />
  <CapturedPokemons
    handleOnClick={[Function]}
    pokemons={Map {}}
  />
</div>
`;
